### PR TITLE
RNN Module

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

Adds the RNN Module according to #2585. This PR includes:
* The `RNN` class
* Updates the `seq2seq` example to use `RNN`
* Exposes `RNNCellBase` in the `linen` module and adds it to the sphinx docs

### CODE REVIEW

File: flax/linen/recurrent.py [hunk 1]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+
+  To support variable length sequences, you can pass a ``segmentation_mask`` which is an integer
+  array of shape ``(*batch, time)`` 1s indicate the elements are part of the sequence and 0s indicate
+  padding elements. Sequences must be padded to the right, i.e. all elements of a sequence must be
+  contiguous and padded elements must be to the right of the sequence. For example::
+
+  TODO: improve example
+
+    >>> # 3 sequences with max length 5
+    >>> segmentation_mask = jnp.array([
```
Comment (marcvanzee) [inline flax/linen/recurrent.py:561]: I forgot the discussion around this, but why don't we just provide a list of sequence lengths? Isn't that more concise? So here we would have `seq_len = jnp.array([3, 2, 5])`.

File: flax/linen/recurrent.py [hunk 2]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+        init_key, batch_dims=batch_dims, size=self.cell_size)
+    else:
+      carry = initial_carry
+
+    def scan_fn(
+      cell: RNNCellBase, carry: Carry, x: Array
+    ) -> Union[Tuple[Carry, Array], Tuple[Carry, Tuple[Carry, Array]]]:
+      carry, y = cell(carry, x)
+
+      if segmentation_mask is None:
```
Comment (marcvanzee) [inline flax/linen/recurrent.py:673]: Could you add a comment why you need this distinction here?

File: flax/linen/recurrent.py [hunk 3]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+      split_rngs=self.split_rngs,
+    )
+
+    scan_output = scan(self.cell, carry, inputs)
+
+    if segmentation_mask is None:
+      carry, outputs = scan_output
+    else:
+      _, (carries, outputs) = scan_output
+      carry = _select_last(carries, segmentation_mask[None], axis=0)
```
Comment (marcvanzee) [inline flax/linen/recurrent.py:695]: Same here, a comment would be useful.

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Let's goooooo! 🚀

File: examples/seq2seq/models.py [hunk 4]
```diff
@@ -189,12 +96,27 @@ def __call__(self, encoder_inputs: Array,
...
     """
     # Encode inputs.
-    init_decoder_state = Encoder(
-        hidden_size=self.hidden_size, eos_id=self.eos_id)(encoder_inputs)
-    # Decode outputs.
-    logits, predictions = Decoder(
-        init_state=init_decoder_state,
-        teacher_force=self.teacher_force,
-        vocab_size=self.vocab_size)(decoder_inputs[:, :-1])
+    encoder = nn.RNN(nn.LSTMCell(), self.hidden_size, return_carry=True, name='encoder')
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply examples/seq2seq/models.py:99]: 🙂

File: flax/linen/recurrent.py [hunk 5]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+    ...
+    >>> x = jnp.ones((10, 50, 32)) # (batch, time, features)
+    >>> lstm = nn.RNN(nn.LSTMCell(), cell_size=64)
+    >>> variables = lstm.init(jax.random.PRNGKey(0), x)
+    >>> y = lstm.apply(variables, x)
+    >>> y.shape # (batch, time, cell_size)
+    (10, 50, 64)
+
+  RNN accepts inputs with shape ``(*batch, time, *features)`` by default, if you set
+  ``time_major=True`` it will instead expect inputs with shape ``(time, *batch, *features)``::
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:532]: I believe `time_major` is a preferred layout preferred for performance on GPUs. 
Couldn't find it in the code base but I *think* (please correct me if I'm wrong) `nn.scan` also converts to time major internally.

File: flax/linen/recurrent.py [hunk 6]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
     key1, key2 = random.split(rng)
     mem_shape = batch_dims + size
     return init_fn(key1, mem_shape), init_fn(key2, mem_shape)
+
+class RNN(Module):
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:514]: ❤️

File: flax/linen/recurrent.py [hunk 7]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+    >>> variables = lstm.init(jax.random.PRNGKey(0), x)
+    >>> carry, y = lstm.apply(variables, x)
+    >>> jax.tree_map(jnp.shape, carry) # ((batch, cell_size), (batch, cell_size))
+    ((10, 64), (10, 64))
+    >>> y.shape # (batch, time, cell_size)
+    (10, 50, 64)
+
+  To support variable length sequences, you can pass a ``segmentation_mask`` which is an integer
+  array of shape ``(*batch, time)`` 1s indicate the elements are part of the sequence and 0s indicate
+  padding elements. Sequences must be padded to the right, i.e. all elements of a sequence must be
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:555]: Thanks! Suggestion copied over.

File: flax/linen/recurrent.py [hunk 8]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+  contiguous and padded elements must be to the right of the sequence. For example::
+
+  TODO: improve example
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:558]: Will remove.

File: flax/linen/recurrent.py (hunk 1 above)
Comment (cgarciae) [reply flax/linen/recurrent.py:561]: This format is compatible with sequence packing so we kept it in case we want that feature in the future. If we want to ditch sequence packing I'd be in favor of the change.

File: flax/linen/recurrent.py [hunk 9]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+
+    >>> # 3 sequences with max length 5
+    >>> segmentation_mask = jnp.array([
+    ...   [1, 1, 1, 0, 0], # length 3
+    ...   [1, 1, 0, 0, 0], # length 2
+    ...   [1, 1, 1, 1, 1], # length 5
+    ... ])
+
+  The output will elements corresponding to padding elements are NOT zeroed out. If ``return_carry``
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:567]: Done.

File: flax/linen/recurrent.py [hunk 10]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+  is set to ``True`` the carry will be the state of the last valid element of each sequence.
+
+  RNN also accepts some of the arguments of :func:`flax.linen.scan`, by default they are set to
+  work with cells like :class:`LSTMCell` and :class:`GRUCell` but you can be overriden as needed.
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:571]: Done

File: flax/linen/recurrent.py [hunk 11]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+  Specifying the default values would look like this::
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:572]: Done

File: flax/linen/recurrent.py [hunk 12]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+      it can be an integer or a tuple of integers.
+    time_major: if ``time_major=False`` (default) it will expect inputs with shape
+      ``(*batch, time, *features)``, else it will expect inputs with shape ``(time, *batch, *features)``.
+    return_carry: if ``return_carry=False`` (default) only the output sequence is returned,
+      else it will return a tuple of the final carry and the output sequence.
+    unroll: how many scan iterations to unroll within a single iteration of a loop,
+      defaults to 1. This argument will be passed to `nn.scan`.
+    variable_axes: the variable collections that are scanned over, this argument will
+      be passed to `nn.scan`.
+    variable_broadcast: specifies the broadcasted variable collections, this argument
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:591]: Improved this a bit by coping over the docstring from `nn.scan` when they were better (not always the case).

File: flax/linen/recurrent.py [hunk 13]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+      will be passed to `nn.scan`.
+    variable_carry: specifies the variable collections that are carried through
+      the loop, this argument will be passed to `nn.scan`.
+    split_rngs: a mapping from PRNGSequenceFilter to bool specifying whether to
+      a PRNG collection should be split such that its values are different at each
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:596]: Done

File: flax/linen/recurrent.py [hunk 14]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+  variable_broadcast: lift.CollectionFilter = 'params'
+  variable_carry: lift.CollectionFilter = False
+  split_rngs: Mapping[lift.PRNGSequenceFilter, bool] = FrozenDict({'params': False})
+
+  def __call__(
+    self,
+    inputs: jax.Array,
+    *,
+    initial_carry: Optional[Carry] = None,
+    init_key: Optional[random.KeyArray] = None,
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:615]: Makes sense in this context, currently all of our cells initialize their carries using `zeros` so they effectively ignore it. Requiring it would be slightly annoying.

File: flax/linen/recurrent.py [hunk 15]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+      return_carry: if ``return_carry=False`` (default) only the output sequence is returned,
+        else it will return a tuple of the final carry and the output sequence.
+      time_major: if ``time_major=False`` (default) it will expect inputs with shape
+        ``(*batch, time, *features)``, else it will expect inputs with shape ``(time, *batch, *features)``.
+    Returns:
+      if ``return_carry=False`` (default) only the output sequence is returned,
+      else it will return a tuple of the final carry and the output sequence.
+    """
+
+    if return_carry is None:
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:644]: Did think about it but logic doesn't match exactly, here we have defaults on the constructor arguments.

File: flax/linen/recurrent.py [hunk 16]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+    if time_major:
+      batch_dims = inputs.shape[1:-num_features_dims]
+    else:
+      batch_dims = inputs.shape[:time_axis]
+
+    carry: Carry
+    if initial_carry is None:
+      if init_key is None:
+        init_key = random.PRNGKey(0)
+      carry = self.cell.initialize_carry(
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:663]: This logic takes care of that:

```python
num_features_dims = 1 if isinstance(self.cell_size, int) else len(self.cell_size)
```

File: flax/linen/recurrent.py [hunk 17]
```diff
@@ -499,3 +510,205 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
...
+        init_key, batch_dims=batch_dims, size=self.cell_size)
+    else:
+      carry = initial_carry
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:666]: A wild walrus has appeared! 🦭

[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/recurrent.py:666]: I know we don't officially support python 3.7 but this will brake them even if we remain compatible for a while 🥲

File: flax/linen/recurrent.py (hunk 2 above)
Comment (cgarciae) [reply flax/linen/recurrent.py:673]: Its a small optimization, if we have no segmentation mask we don't need to have the carry as an output. The reason we have the carry as an output is because its faster to select the last carry out side the scan loop than have an `jnp.where` on each step.

I think we should expand this to:

```python
if segmentation_mask is None or not return_carry:
```

File: flax/linen/recurrent.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/recurrent.py:657]: But doesn't that still work? It will choose the value that is not `None`, which in this case would be the constructor argument.

File: flax/linen/recurrent.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/recurrent.py:676]: Could you then add a test where `cell_size` is a tuple rather than an int to make sure it works? (I think right now you only have tests where it is an integer equal to `channels_out`.

File: flax/linen/recurrent.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/recurrent.py:679]: Ok, I'm fine either way (let's agree on a policy with the team)

**Comment by marcvanzee** [comment]
Also the BUILD is failing. could you please look into that as well?

File: flax/linen/recurrent.py (hunk 1 above)
Comment (cgarciae) [reply flax/linen/recurrent.py:571]: Comment added.

File: flax/linen/recurrent.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/recurrent.py:676]: Seems you have this test already, so please ignore.

File: flax/linen/recurrent.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/recurrent.py:657]: After discussing offline, it seems merge_param is xor, which is not what you want here. Resolving.

**Comment by cgarciae** [comment]
@marcvanzee I believe this is ready :)
